### PR TITLE
State Channel example contract

### PR DESCRIPTION
**Pull request by velzevur**

### CODE REVIEW

File: sophia.tex [hunk 1]
```diff
@@ -155,146 +155,187 @@ \subsubsection{Dutch auction contract}
...
+participants in their own context. This requires both to share the same view
+of the chain\footnote{We achieve this by specifying the on-chain environment
+as part of the off-chain update.}.
+
+Running a simple insurance contract in which one user provides the service
+while the other insures their property for a certain timeframe from hailstorm
+requires some input from real life data. This is where oracle input comes: for
+the example we have an on-chain Oracle to produce data if there had been a
+hailstorm in a certain city. It publishes data pinned in time with a
+corresponding block height when it occurred. Later on insured participant can
```
Comment (uwiger) [inline sophia.tex]:
```suggestion
corresponding block height when it occurred. Later on insured participants can
```

File: sophia.tex [hunk 2]
```diff
@@ -155,146 +155,187 @@ \subsubsection{Dutch auction contract}
...
+Contracts can be deployed on-chain or in a state channel. State channels
+provide means for two parties for executing transactions off-chain. Once the
+channel is opened on-chain all the communication is moved off-chain and it is
+just between the two participants. It is not limited by blockchain
+throughputh, fees or gas costs. This provides a new environment to run
+contracts in. In order to make most out of it, off-chain contracts can use
+on-chain data.
+
+An example of the use of off-chain contract and an on-chain oracle can be
+illustrated with a little insurance example. Note that all computation on a
```
Comment (uwiger) [inline sophia.tex]:
```suggestion
illustrated with a simple insurance contract. Note that all computation on a
```

File: sophia.tex [hunk 3]
```diff
@@ -155,146 +155,187 @@ \subsubsection{Dutch auction contract}
...
+If not exactly the same, the corresponding state hashes will differ. This is
+true especially with state channels as contracts are executed by both
+participants in their own context. This requires both to share the same view
+of the chain\footnote{We achieve this by specifying the on-chain environment
+as part of the off-chain update.}.
+
+Running a simple insurance contract in which one user provides the service
+while the other insures their property for a certain timeframe from hailstorm
+requires some input from real life data. This is where oracle input comes: for
+the example we have an on-chain Oracle to produce data if there had been a
```
Comment (uwiger) [inline sophia.tex]:
```suggestion
the example we have an on-chain Oracle to produce data on whether there has been a
```

File: sophia.tex [hunk 4]
```diff
@@ -155,146 +155,187 @@ \subsubsection{Dutch auction contract}
...
-because initially there is no ongoing lottery. The state also contains
-an oracle and an optional query. This query will be instantiated when
-the lottery is closed and a ticket is drawn.
-
-The creator of this contract may now start the lottery by supplying a relative
-closing height,
-for example, 20 key-blocks from that the transaction gets on chain;
-approximately one hour. Any user can then buy a ticket.
+The contract is deployed by the insurance service provider in a state channel.
+It stores the trusted oracle in its state, the price it expects for insurance
```
Comment (uwiger) [inline sophia.tex]: stores a reference to?

File: sophia.tex [hunk 5]
```diff
@@ -155,146 +155,187 @@ \subsubsection{Dutch auction contract}
...
+not expected to call that function - any attempt of doing so will result in
+aborting the call. This is true both off-chain and on-chain in a forced
+progress scenario.
+
+The keyword \verb+payable+ expresses that we expect the caller to add a token
+amount to the contract call transaction, which could be checked by comparing
+\verb+Call.value+. There is no restriction for the amount being deposited in
+the contract and this could be done in a couple of different steps if needed.
+
+Once that the compensation is present in the contract, the other party can
```
Comment (uwiger) [inline sophia.tex]:
```suggestion
Once the compensation is present in the contract, the other party can
```

File: sophia.tex [hunk 6]
```diff
@@ -155,146 +155,187 @@ \subsubsection{Dutch auction contract}
...
-is not around. The TTLs in the query assure that the oracle has
-approximately 15 minutes to answer, long enough to even get it out
-in busy times. Within a day one then has to claim the price sum,
-otherwise, because the query answer is then no longer on chain.
+There are some checks that it is not the insurance provider that is calling
+the contract and that the contract has enough tokens to cover the compensation
+in case of a hailstorm. If those pass the corresponding insurance period is
+calculated according to the tokens dedicated in the contract and the state is
+updated accordingly. Note that if there is an active insurance at the moment,
+existing time frame is just extended. This allows for continuous insurances.
```
Comment (uwiger) [inline sophia.tex]:
```suggestion
the existing time frame is just extended. This allows for continuous insurances.
```

File: sophia.tex [hunk 7]
```diff
@@ -155,146 +155,187 @@ \subsubsection{Dutch auction contract}
...
+        require(city == state.city, "different_city")
+        let is_in_range = generation >= state.active_from && generation =< state.active_to
+        require(is_in_range, "not_in_insurance_range")
+        require(was_there_hailstorm, "different_response")
+        Chain.spend(Call.caller, state.compensation)
+        put(state{ active_from = 0,
+                   active_to = 0})
+\end{verbatim}
+
+Note that if any of the check fails, the whole execution of the contract
```
Comment (uwiger) [inline sophia.tex]:
```suggestion
Note that if any of the checks fail, the whole execution of the contract
```

File: sophia.tex (hunk 1 above)
Comment (velzevur) [reply sophia.tex]: Well one is insuring the other, so only one is being insured 🤔

Comment (uwiger) [reply sophia.tex]: Ok, then it should be "the insured participant"

**Review by uwiger — APPROVED** [review]

File: sophia.tex (hunk 1 above)
Comment (velzevur) [reply sophia.tex]: Addressed

**Comment by velzevur** [comment]
Addressed all comments, regenerated the pdf document and squashed changes in a single commit.

**Comment by velzevur** [comment]
Rebased

**Review by keypair — APPROVED** [review]

**Pull request merged**